src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/JavaSourceBuilder.java

```diff
@@ -259,12 +203,74 @@ protected String addressGetCallString(String javaName, String nativeName, Memory
      * Header$CFooS, Header$CfooS, Header$CFoOs and so on! We solve this by
      * generating unique case-insensitive names for nested classes.
      */
-    String uniqueNestedClassName(String name) {
+    final String uniqueNestedClassName(String name) {
         name = Utils.javaSafeIdentifier(name);
         return nestedClassNames.add(name.toLowerCase()) ? name : (name + "$" + nestedClassNameCount++);
     }
 
-    StructBuilder newStructBuilder(String name, GroupLayout parentLayout, Type type) {
-        return new StructBuilder(this, name, parentLayout, type);
+    protected void emitPackagePrefix() {
+        assert packageName().indexOf('/') == -1 : "package name invalid: " + packageName();
+        append("// Generated by jextract\n\n");
+        if (!packageName().isEmpty()) {
+            append("package ");
+            append(packageName());
+            append(";\n\n");
+        }
+    }
+
+    protected void emitImportSection() {
+        append("import java.lang.invoke.MethodHandle;\n");
+        append("import java.lang.invoke.VarHandle;\n");
+        append("import java.util.Objects;\n");
+        append("import jdk.incubator.foreign.*;\n");
+        append("import jdk.incubator.foreign.MemoryLayout.PathElement;\n");
+        append("import static ");
+        append(OutputFactory.C_LANG_CONSTANTS_HOLDER);
+        append(".*;\n");
+    }
+
+    protected void emitGetter(String mods, Class<?> type, String name, String access, boolean nullCheck, String errMsg) {
+        incrAlign();
+        indent();
+        append(mods + " " + type.getSimpleName() + " " +name + "() {\n");
+        incrAlign();
+        indent();
+        append("return ");
+        if (nullCheck) {
+            append("RuntimeHelper.requireNonNull(");
+        }
+        append(access);
+        if (nullCheck) {
+            append(",\"");
+            append(errMsg);
+            append("\")");
+        }
+        append(";\n");
+        decrAlign();
+        indent();
+        append("}\n");
+        decrAlign();
+    }
+
+    protected void emitGetter(String mods, Class<?> type, String name, String access) {
+        emitGetter(mods, type, name, access, false, null);
+    }
+
+    int constant_counter = 0;
+
+    static final int CONSTANTS_PER_CLASS = Integer.getInteger("jextract.constants.per.class", 5);
+    ConstantBuilder constantBuilder;
+
+    protected void emitWithConstantClass(String javaName, Consumer<ConstantBuilder> constantConsumer) {
+        if (constant_counter > CONSTANTS_PER_CLASS || constantBuilder == null) {
+            if (constantBuilder != null) {
+                constantBuilder.classEnd();
+            }
+            constant_counter = 0;
+            constantBuilder = new ConstantBuilder(this, Kind.CLASS, javaName + "_constants");
```

Comment:
Using the java name of some constant to uniqueify the name of the constant class seems a bit strange to me... Couldn't there be collisions? (e.g. if struct and unction are named the same?) Why not just use to id/number here like before?

Reply:
there could be collisions, yes - not sure I get what you propose by using "id/number"

Reply:
I see what you mean - I thought you meant collisions in case an header contains `foo` and `foo_constants` - but you mean conflicts between members that have same native name.

Reply:
Yes. With id/number I mean what [MultiFileConstantHelper::getConstantClassName()](https://github.com/openjdk/panama-foreign/pull/450/files#diff-0c06a52b2e1c4375f51e101bbbd60d6dc8405e9ace3d6d6a8608e2b184146263L62) was doing before, i.e. using the number of constant classes emitted so far as a suffix id.